### ISSUE TITLE
graph: non-weighted edge representation

### ISSUE THREAD

**Issue by mewmew** [gonum/gonum#203]
This issue opens up for a general discussion on the edge representation used in gonum/graph. Currently the graph.Edge interface requires a `Weight` method, which is required for some applications (e.g. shortest path with different costs between nodes) but stubbed out with a dummy implementation for others (e.g. control flow graphs and call graphs).

The intention is to investigate whether it would be preferable to keep the edge representation minimal, and let it be extended to include notions of weights. For instance:

```go
// Edge is a graph edge. In directed graphs, the direction of the edge is given from -> to, otherwise the edge is semantically unordered. 
type Edge interface {
    From() Node
    To() Node
}

// WeightedEdge is a graph edge with an associated weight.
type WeightedEdge interface {
    Edge
    // Weight returns the weight of the edge.
    Weight() float64
}
```

An initial discussion has been started in #201, relevant comment extracts included below.

From @mewmew in https://github.com/gonum/gonum/pull/201#issuecomment-323330362

>> The makes use of the new graph.EdgeAdder interface (currently the functionality is degraded because we cannot set the weight here - though do we want to?).
>
> This is something I've been meaning to bring up in the past, to discuss whether an edge should be required to have a weight, or if simply there should be edges which implement a Weighter interface or something along those lines.
>
> The rationale for not including weights in the minimal edge representation is that there are many kinds of graphs which don't make use of weights, and thus are required to include dummy values or dummy methods for edges. This happens to be the case for the majority of my personal applications of graphs, namely control flow graphs and call graphs.

From @kortschak in https://github.com/gonum/gonum/pull/201#issuecomment-323471659:

> The issue around weights is difficult. The discussion happened in PRs and issues in gonum/graph if you want to wade through that. Briefly though, in order to be able to distinguish between weighted and unweighted graphs we need to have method set differences between the graph types (not the edge types, since the caller doesn't necessarily see those).
>
> I guess it would be possible to change this so that edges did not by default have weights, but weighted graphs return a type WeighedEdge interface { Edge; Weight() float64 } from a WeightedEdgeBetween(x, y Node) WeightedEdge method etc. This increases the burden on implementers to add those additional methods (though they are trivially just the base EdgeBetween method with a different return type (and the base method then just wraps them with the Edge return type.
>
> I think I would be OK with this. Thoughts?

**Comment by mewmew** [comment]
> I guess it would be possible to change this so that edges did not by default have weights, but weighted graphs return a type WeighedEdge interface { Edge; Weight() float64 } from a WeightedEdgeBetween(x, y Node) WeightedEdge method etc. ...
>
> I think I would be OK with this. Thoughts?

I would also wish to explore this path, separating `Edge` from the `WeightedEdge` representation seems reasonable and I think we should try to arrive there somehow. Exactly if adding a `WeightedEdgeBetween` method is the right approach or if there are others we may wish to evaluate, I still have to give some thought.

What happens if a graph has an edge between `x` and `y`, but the edge is not a weighted edge, and `WeightedEdgeBetween(x, y)` is invoked. What would it return? `nil`? or the unweighted edge but with a dummy weight, e.g. 0?

**Comment by kortschak** [comment]
My thinking at the moment is that we add the following types (and maybe remove `Weighter`):

```
type WeightedGraph interface {
    Graph

    // HasEdgeBetween returns whether an edge exists between
    // nodes x and y without considering direction.
    HasWeightedEdgeBetween(x, y Node) bool

    // Edge returns the edge from u to v if such an edge
    // exists and nil otherwise. The node v must be directly
    // reachable from u as defined by the From method.
    WeightedEdge(u, v Node) WeightedEdge

    // Weight returns the weight for the edge between
    // x and y if Edge(x, y) returns a non-nil Edge.
    // If x and y are the same node or there is no
    // joining edge between the two nodes the weight
    // value returned is implementation dependent.
    // Weight returns true if an edge exists between
    // x and y or if x and y have the same ID, false
    // otherwise.
    Weight(x, y Node) (w float64, ok bool)
}
```

with similar types for `WeightedUndirected` and `WeightedDirected`.

**Comment by mewmew** [comment]
Let `g` be of a type implementing `WeightedGraph`, would `g.HasEdgeBetween` be a synonym for `g.HasWeightedEdgeBetween`?

Their comments and function signatures are identical (except for the method name).

The `WeightedEdge` method makes sense, as it would basically be a convenience function for returning edges which implement the `WeightedEdge` interface, thus letting users skip a type assertion.

In general I agree with your approach of defining a WeightedGraph interface (and also getting rid of the `Weighter` interface in the process), as algorithms which require weights to function should take `WeightedGraph`s as input, rather than simply `Graph`. This would improve the API contracts, as the intention is signalled through type-checked code rather than comments.

**Comment by kortschak** [comment]
Whether they are synonymous depends on whether a graph can be mixed weighted and unweighted. I can't see any good reason for one to be. If this is the case, then they would be synonymous and the weighted case can be dropped. If there is a reason for mixed weighting, then they would not be synonymous

I think dropping it is there right way forward.

**Comment by kortschak** [comment]
The obvious corollary here is that your add-edge branch should be merged and another WeightedEdgeAdder type should be added.

**Comment by mewmew** [comment]
> I think dropping it is there right way forward.

I agree.

> The obvious corollary here is that your add-edge branch should be merged and another WeightedEdgeAdder type should be added.

Indeed, this affects both read (i.e. `Graph` and `WeightedGraph`) and write (i.e. `EdgeAdder` and `WeightedEdgeAdder`).

**Comment by kortschak** [comment]
This was closed by #206.
